First version of retrieval tutorial (#1)
nasa-fornax/fornax-howtos: tutorials/exoplanet_retrievals.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.19.1
kernelspec:
  name: python3
  display_name: python3
  language: python
---

> # Exoplanet Retrievals

Reviewer: > Title preference: should this how-to emphasize the science use case (“Exoplanet Retrievals”) or the workflow skills (“Creating Environments and Running Scripts”)?

I'm inclined towards emphasizing the workflow skills since it's going in this "how to" repo rather than in the science showcase demos. But I see that the notebook is much more focused on the specific setup, etc. needed for exoplanet retrievals, and I feel the title should match. So now I've changed my mind. I'd suggest "Exoplanet Retrieval Quickstart" or something similar to indicate that it's not a fully worked science demo.

Author: I renamed this file to be quickstart.  I will also open an issue to make a "howto" on just the making a conda environment part of this - which we can lift from here as soon as this finishes its code review - and add some verbiage around.

## Learning goals

By the end of this tutorial, you will be able to:

1. Set up a persistent environment for retrieval software on Fornax.
2. Install and activate dependencies for pRT retrieval workflows.
3. Run Python retrieval scripts with `mpirun`for optimal parallelization.
4. Benchmark scaling behavior to estimate realistic runtime.

## Introduction

> This tutorial shows how to run an exoplanet atmospheric retrieval workflow on the Fornax Science Console, using [petitRADTRANS](https://petitradtrans.readthedocs.io/en/latest/index.html) (pRT) + [`pymultinest`](https://johannesbuchner.github.io/PyMultiNest/) + [Message Passing Interface](https://www.mpich.org) (MPI) as an example stack. This code follows the example basic retrieval [tutorial](https://petitradtrans.readthedocs.io/en/latest/content/notebooks/retrieval_basic.html) from pRT. This tutorial is intended to show the steps involved in getting exoplanet retrievals to run on a system like Fornax, and does not show the details and specifics of how to tailor existing retrieval codes to match specific datasets.

Author: ```suggestion
This tutorial shows how to run an exoplanet atmospheric retrieval workflow on the Fornax Science Console, using [petitRADTRANS](https://petitradtrans.readthedocs.io/en/latest/index.html) (pRT) + [`pymultinest`](https://johannesbuchner.github.io/PyMultiNest/) + [Message Passing Interface](https://www.mpich.org) (MPI) as an example stack. 
This code follows the example basic retrieval [tutorial](https://petitradtrans.readthedocs.io/en/latest/content/notebooks/retrieval_basic.html) from pRT. 
This tutorial is intended to show the steps involved in getting exoplanet retrievals to run on a system like Fornax, and does not show the details and specifics of how to tailor existing retrieval codes to match specific datasets.
```

## Instructions

> Use this guide along with a terminal to setup and run basic exoplanet retrievals.

Reviewer: A little more text would be helpful to explain that the notebook itself doesn't execute anything; users have to copy/paste commands into a terminal. I'm not sure users will understand that just from "Use this guide along with a terminal".

> Use this guide along with a terminal to setup and run basic exoplanet retrievals.

Reviewer: A little more text would be helpful to explain that the notebook itself doesn't execute anything; users have to copy/paste commands into a terminal. I'm not sure users will understand that just from "Use this guide along with a terminal".

Author: done.

> Use this guide along with a terminal to setup and run basic exoplanet retrievals.

Author: ```suggestion
Use this guide as a companion reference while working in a Fornax terminal session.

```{admonition} Important
This notebook does **not** execute any of the setup or MPI commands shown below.  
All commands shown in code blocks must be run manually in a terminal by copying and pasting them. 
The notebook serves only as documentation of the workflow.
```

+++

## 1. Create a persistent environment for retrieval dependencies

> Retrieval workflows generally require tightly constrained scientific software stacks. On Fornax, use a persistent user-managed conda/micromamba environment.

Reviewer: Include a link to our docs page that has the instructions. https://docs.fornax.sciencecloud.nasa.gov/compute-environments/

Author: done.

> Retrieval workflows generally require tightly constrained scientific software stacks. On Fornax, use a persistent user-managed conda/micromamba environment.

Author: ```suggestion
Retrieval workflows generally require tightly constrained scientific software stacks. 
On Fornax, use a persistent user-managed conda/micromamba environment.
Instructions for how to do that are here: https://docs.fornax.sciencecloud.nasa.gov/compute-environments/
```

> Create a file named `conda-exo.yml` with the following contents.  One is also provided in the examples/exoplanet_retrievals directory for convenience.  Note that the name of the file must start with "conda-" for the setup script to find it.

Author: ```suggestion
Create a file named `conda-exo.yml` with the following contents.  
One is also provided in the examples/exoplanet_retrievals directory for convenience.  
Note that the name of the file must start with "conda-" for the setup script to find it.
```

```yaml
name: exo
channels:
  - conda-forge
dependencies:
  - python=3.12
  - pymultinest
  - c-compiler
  - cxx-compiler
  - fortran-compiler
  - pip
  - pip:
      - matplotlib
      - numba
      - "petitRADTRANS[retrieval]"
      - pandas
```

### 1.1 Build the environment:

From the terminal, build the environment with the following command:
```bash
setup-conda-env --user conda-exo.yml

> ```

Reviewer: Does this work for you? I'm pretty sure that script won't take a file name as input. Based on my recent experience with the spherex source discovery tool, I think it needs to be like this:

````suggestion
From the terminal, `cd` into the directory where you saved the conda file, then build the environment with the following command:

```bash
setup-conda-env --user
```
````

Notes:

- The tool prompts you to confirm the environment name.
- First-time environment creation may take more than 10 minutes.

> - Retrieval stacks can consume significant storage.

Reviewer: What are "Retrieval stacks"? It sounds like output from running code and not anything to do with building the environment. In that case, move that line to a different section.

Author: done.

> - Retrieval stacks can consume significant storage.

Author: ```suggestion
- Environments containing packages to run retrievals can consume significant storage.
```

### 1.2 Activate the environment:

From the terminal, activate the environment with the following command:
```bash
micromamba activate $USER_ENV_DIR/exo
```

+++

## 2. Install opacity files (required and large)

Many retrieval tools (including pRT) require local opacity files that are too large to package in this repository.

- You must download/install opacities in your own Fornax storage.
- Opacity files are **not** distributed in this repo due to size constraints.
- pRT opacity installation instructions: <https://petitradtrans.readthedocs.io/en/latest/content/available_opacities.html>

If centralized opacity hosting on Fornax would help your workflow, please share feedback with the Fornax team.

+++

## 3. Run retrieval scripts with MPI

For pRT workflows that use `pymultinest`, write retrieval wrappers as Python scripts and launch with `mpirun`.

This repository includes:

- `examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py`
- `examples/exoplanet-retrievals/hst_example_clear_spec.txt`

Example MPI launch commands:

> 

Author: ```suggestion
Note: The --n-live-points argument sets the number of “live” samples (n_live in MultiNest terminology) maintained during nested sampling; larger values improve the stability of evidence and posterior estimates but increase runtime.
```

```bash
# Small validation run (recommended first test)
mpirun -np 2 python examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py --use-mpi --n-live-points 40

# Medium run for quick scaling checks (assuming you have ncores > 8)
mpirun -np 8 python examples/exoplanet-retrievals/run_prt_basic_pymultinest_mpi.py --use-mpi --n-live-points 40
```

### Important Fornax MPI subtleties

> When running pRT retrievals with MPI on Fornax, using all available CPU cores can exhaust limited shared-memory space (`/dev/shm`) used for inter-process communication. This can fail jobs even when system RAM is still available.

Author: ```suggestion
When running pRT retrievals with MPI on Fornax, using all available CPU cores can exhaust limited shared-memory space (`/dev/shm`) used for inter-process communication. 
This can fail jobs even when system RAM is still available.
```

In practice:

1. Speedup is usually sub-linear.
2. Performance gains often flatten before you reach all available cores.
3. Many workflows perform best around roughly half to three-quarters of available cores.

+++

## 4. Benchmark to estimate full retrieval runtime

> Full retrieval wall time is difficult to predict in advance. Benchmarking your own setup (opacities, chemistry assumptions, cloud setup, live points) helps you choose compute size and decide whether you need keep-alive sessions for runs longer than 24 hours.

Reviewer: Point to our docs that explain what "keep-alive" is. https://docs.fornax.sciencecloud.nasa.gov/jupyterlab/#jupyterlab-session-information

Also, what's the significance of 24 hours here? I'm not aware of anything fornax related that triggers at 24 hours. Same question for the last bullet in the "Recommended workflow checklist" section below.

Author: I had thought that all sessions were culled after 24 hours.  I see now that the docs say 48 hours.  I also just don't trust the culling activity so if I were going to run a job that I knew took a long time, I would use the stay alive to be safe.  I've tried to capture this in the new edit.

Reviewer: Servers are subject to being culled after 15 minutes of no CPU activity. They're not guaranteed to be culled at exactly 15 minutes but the user shouldn't count on anything longer than that. 48 hours is the hard upper limit. Even keep-alive won't keep a server running longer then that.

> Full retrieval wall time is difficult to predict in advance. Benchmarking your own setup (opacities, chemistry assumptions, cloud setup, live points) helps you choose compute size and decide whether you need keep-alive sessions for runs longer than 24 hours.

Author: ```suggestion
Full retrieval wall time is difficult to predict in advance. 
Benchmarking your own setup (opacities, chemistry assumptions, cloud setup, live points) helps you choose compute size and estimate total runtime. 
If you anticipate that your retrieval will run longer than a few hours, enabling the [JupyterLab keep-alive feature](https://docs.fornax.sciencecloud.nasa.gov/jupyterlab/#jupyterlab-session-information) can prevent unintended shutdown. 
However, keep-alive sessions incur credit charges for their entire lifetime, regardless of active CPU usage, so plan accordingly and monitor extended runs.
```

### Benchmark figure 1: speedup vs nprocs

```{figure} ../examples/exoplanet-retrievals/speedup_vs_nprocs.png
:alt: Speedup versus number of MPI processes
:name: fig-speedup-vs-nprocs

> Speedup is defined as the wall time for 1 processor divided by the wall time for *nprocs* processors. Typical behavior includes (1) non-linear scaling, (2) useful speedup up to about 0.5-0.75× of available cores, and (3) larger gains for larger `n_live_points`, which is more representative of realistic retrieval configurations. Overheads are a major cause of non-linear speedups and limits in gains as a function of `nprocs`.

Reviewer: Describe what `n_live_points` and "n_live" (legend) are.

Author: done.  here and higher up.

> Speedup is defined as the wall time for 1 processor divided by the wall time for *nprocs* processors. Typical behavior includes (1) non-linear scaling, (2) useful speedup up to about 0.5-0.75× of available cores, and (3) larger gains for larger `n_live_points`, which is more representative of realistic retrieval configurations. Overheads are a major cause of non-linear speedups and limits in gains as a function of `nprocs`.

Author: ```suggestion
Speedup is defined as the wall time for 1 processor divided by the wall time for *nprocs* processors.  
Typical behavior includes (1) non-linear scaling, (2) useful speedup up to about 0.5-0.75× of available cores, and (3) Larger gains are seen for larger n_live_points (shown as n_live in the legend), which controls the number of active samples MultiNest keeps while exploring parameter space. 
Increasing n_live_points improves posterior and evidence stability but increases runtime.
Overheads are a major cause of non-linear speedups and limits in gains as a function of `nprocs`.
```

```

### Benchmark figure 2: seconds per likelihood evaluation vs nprocs

```{figure} ../examples/exoplanet-retrievals/seconds_per_likelihood_eval_vs_nprocs.png
:alt: Seconds per likelihood evaluation versus number of MPI processes
:name: fig-seconds-per-like

> A benchmarking curve like this for your specific retrieval setup (including your opacity selection, clouds, and model complexity) can help predict full runtime and guide resource selection.

Reviewer: I'm not sure what this means. Is there something useful that can be gleaned from the example figure or does a user really have to make their own figure?

Author: I have a benchmarking wrapper that runs around the pRT wrapper included here, but I thought that was too much tangential information to include in this tutorial.  My benchmarking wrapper is what generated these plots.  Do you think this tutorial would benefit from that wrapper?

Reviewer: It sounds useful to provide if you already have it, but you're in a much better position to judge whether it's better to leave out than I am. Mostly I was confused by the languaging because the caption for the figure above drew some general conclusions based on the provided figure while this one just said the user should make their own version. So it seemed like the message was that the user should not try to draw any conclusions about this one before making their own version. I see that you've updated this caption and it resolves my confusion.

> A benchmarking curve like this for your specific retrieval setup (including your opacity selection, clouds, and model complexity) can help predict full runtime and guide resource selection.

Author: ```suggestion
The example figure illustrates typical scaling behavior for this configuration, but runtime depends strongly on your specific retrieval setup. 
To obtain a reliable estimate for your own science case, repeat short test runs at multiple -np values (e.g., 2, 4, 8, etc.) and a range of n_live_points, record the wall time for each, and construct a similar scaling curve. 
This allows you to estimate total runtime and select compute resources based on measured performance rather than core count alone.
```

```

+++

## 5. Recommended workflow checklist

1. Build and activate your persistent retrieval environment.
2. Install pRT opacities in persistent user storage.
3. Validate script execution at low `-np` first.
4. Benchmark multiple `-np` values.

> 5. Choose a resource size based on measured scaling, not core count alone.

Reviewer: What's a "resource size"? Does it mean server size?

Author: changed.

> 5. Choose a resource size based on measured scaling, not core count alone.

Author: ```suggestion
5. Choose a server size based on measured scaling, not core count alone.
```
